Linux amalgamation builds fail to link `libdl.so`

**What does happen?**
Amalgamation builds of both the latest release (0.2.7) and HEAD fail as follows:
```
$ mkdir -p build/amalgamation
$ python3 scripts/amalgamation.py
-----------------------
-- Writing src/amalgamation/duckdb.hpp --
-----------------------
------------------------
-- Writing src/amalgamation/duckdb.cpp --
------------------------
$ cd build/amalgamation
$ cmake ../.. -DAMALGAMATION_BUILD=ON
-- The C compiler identification is GNU 8.3.0
-- The CXX compiler identification is GNU 8.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Setting build type to 'Release'.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/aho/tmp/duckdb/duckdb/build/amalgamation
$ make
[  2%] Building CXX object src/CMakeFiles/duckdb_static.dir/amalgamation/duckdb.cpp.o
[  5%] Linking CXX static library libduckdb_static.a
[  5%] Built target duckdb_static
[  7%] Building CXX object src/CMakeFiles/duckdb.dir/amalgamation/duckdb.cpp.o
[ 10%] Linking CXX shared library libduckdb.so
[ 10%] Built target duckdb
[ 12%] Building CXX object tools/sqlite3_api_wrapper/sqlite3_udf_api/CMakeFiles/sqlite3_udf_api.dir/sqlite3_udf_wrapper.cpp.o
[ 15%] Building CXX object tools/sqlite3_api_wrapper/sqlite3_udf_api/CMakeFiles/sqlite3_udf_api.dir/cast_sqlite.cpp.o
[ 15%] Built target sqlite3_udf_api
[ 17%] Building C object tools/sqlite3_api_wrapper/sqlite3/CMakeFiles/sqlite3_api_wrapper_sqlite3.dir/printf.c.o
[ 20%] Building C object tools/sqlite3_api_wrapper/sqlite3/CMakeFiles/sqlite3_api_wrapper_sqlite3.dir/strglob.c.o
[ 20%] Built target sqlite3_api_wrapper_sqlite3
[ 22%] Building CXX object tools/sqlite3_api_wrapper/CMakeFiles/sqlite3_api_wrapper.dir/sqlite3_api_wrapper.cpp.o
[ 25%] Linking CXX shared library libsqlite3_api_wrapper.so
[ 25%] Built target sqlite3_api_wrapper
[ 27%] Building CXX object tools/sqlite3_api_wrapper/CMakeFiles/sqlite3_api_wrapper_static.dir/sqlite3_api_wrapper.cpp.o
[ 30%] Linking CXX static library libsqlite3_api_wrapper_static.a
[ 30%] Built target sqlite3_api_wrapper_static
[ 32%] Building CXX object tools/sqlite3_api_wrapper/CMakeFiles/test_sqlite3_api_wrapper.dir/test/test_sqlite3_api_wrapper.cpp.o
[ 35%] Building CXX object tools/sqlite3_api_wrapper/CMakeFiles/test_sqlite3_api_wrapper.dir/test/test_sqlite3_udf_api_wrapper.cpp.o
[ 37%] Linking CXX executable test_sqlite3_api_wrapper
/usr/bin/ld: ../../src/libduckdb.so: undefined reference to `dlopen'
/usr/bin/ld: ../../src/libduckdb.so: undefined reference to `dlsym'
collect2: error: ld returned 1 exit status
make[2]: *** [tools/sqlite3_api_wrapper/CMakeFiles/test_sqlite3_api_wrapper.dir/build.make:115: tools/sqlite3_api_wrapper/test_sqlite3_api_wrapper] Error 1
make[1]: *** [CMakeFiles/Makefile2:683: tools/sqlite3_api_wrapper/CMakeFiles/test_sqlite3_api_wrapper.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
while the corresponding full-source builds (i.e. without `-DAMALGAMATION_BUILD=ON`) complete successfully:
```
$ mkdir -p build/fullsrc
$ cd build/fullsrc
$ cmake ../..
-- The C compiler identification is GNU 8.3.0
-- The CXX compiler identification is GNU 8.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Setting build type to 'Release'.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/aho/tmp/duckdb/duckdb/build/fullsrc
$ make
[  1%] Building CXX object third_party/hyperloglog/CMakeFiles/hyperloglog.dir/hyperloglog.cpp.o
[  2%] Building CXX object third_party/hyperloglog/CMakeFiles/hyperloglog.dir/sds.cpp.o
[  2%] Linking CXX static library libhyperloglog.a
[  2%] Built target hyperloglog
[  2%] Building CXX object src/optimizer/CMakeFiles/duckdb_optimizer.dir/ub_duckdb_optimizer.cpp.o
[  2%] Built target duckdb_optimizer
[  2%] Building CXX object src/optimizer/matcher/CMakeFiles/duckdb_optimizer_matcher.dir/ub_duckdb_optimizer_matcher.cpp.o
[  2%] Built target duckdb_optimizer_matcher
...
[ 84%] Built target sqlite3_api_wrapper_static
[ 84%] Building CXX object tools/sqlite3_api_wrapper/CMakeFiles/test_sqlite3_api_wrapper.dir/test/test_sqlite3_api_wrapper.cpp.o
[ 85%] Building CXX object tools/sqlite3_api_wrapper/CMakeFiles/test_sqlite3_api_wrapper.dir/test/test_sqlite3_udf_api_wrapper.cpp.o
[ 86%] Linking CXX executable test_sqlite3_api_wrapper
# ...amalgamation build fails here...
[ 86%] Built target test_sqlite3_api_wrapper
[ 87%] Building C object tools/shell/CMakeFiles/shell.dir/shell.c.o
[ 87%] Building CXX object tools/shell/CMakeFiles/shell.dir/linenoise.cpp.o
[ 88%] Linking CXX executable ../../duckdb
...
[ 99%] Building C object third_party/sqlite/CMakeFiles/sqlite3.dir/sqlite3.c.o
[100%] Linking C static library libsqlite3.a
[100%] Built target sqlite3
```

A diff of `duckdb.dir/link.txt` from both builds reveals the issue:
```
$ diff -u build/*/src/CMakeFiles/duckdb.dir/link.txt 
--- build/amalgamation/src/CMakeFiles/duckdb.dir/link.txt	2021-07-26 13:55:10.848774782 +0800
+++ build/fullsrc/src/CMakeFiles/duckdb.dir/link.txt	2021-07-26 14:05:34.307686772 +0800
@@ -1 +1 @@
-/usr/bin/c++ -fPIC -O3 -DNDEBUG -O3 -DNDEBUG   -L/brew/lib -shared -Wl,-soname,libduckdb.so -o libduckdb.so CMakeFiles/duckdb.dir/amalgamation/duckdb.cpp.o  -lpthread 
+/usr/bin/c++ -fPIC -O3 -DNDEBUG -O3 -DNDEBUG   -L/brew/lib -shared -Wl,-soname,libduckdb.so -o libduckdb.so optimizer/matcher/CMakeFiles/duckdb_optimizer_matcher.dir/ub_duckdb_optimizer_matcher.cpp.o optimizer/join_order/CMakeFiles/duckdb_optimizer_join_order.dir/ub_duckdb_optimizer_join_order.cpp.o optimizer/pushdown/CMakeFiles/duckdb_optimizer_pushdown.dir/ub_duckdb_optimizer_pushdown.cpp.o optimizer/pullup/CMakeFiles/duckdb_optimizer_pullup.dir/ub_duckdb_optimizer_pullup.cpp.o optimizer/rule/CMakeFiles/duckdb_optimizer_rules.dir/ub_duckdb_optimizer_rules.cpp.o optimizer/statistics/expression/CMakeFiles/duckdb_optimizer_statistics_expr.dir/ub_duckdb_optimizer_statistics_expr.cpp.o optimizer/statistics/operator/CMakeFiles/duckdb_optimizer_statistics_op.dir/ub_duckdb_optimizer_statistics_op.cpp.o optimizer/CMakeFiles/duckdb_optimizer.dir/ub_duckdb_optimizer.cpp.o planner/expression/CMakeFiles/duckdb_planner_expression.dir/ub_duckdb_planner_expression.cpp.o planner/binder/expression/CMakeFiles/duckdb_bind_expression.dir/ub_duckdb_bind_expression.cpp.o planner/binder/query_node/CMakeFiles/duckdb_bind_query_node.dir/ub_duckdb_bind_query_node.cpp.o planner/binder/statement/CMakeFiles/duckdb_bind_statement.dir/ub_duckdb_bind_statement.cpp.o planner/binder/tableref/CMakeFiles/duckdb_bind_tableref.dir/ub_duckdb_bind_tableref.cpp.o planner/expression_binder/CMakeFiles/duckdb_expression_binders.dir/ub_duckdb_expression_binders.cpp.o planner/filter/CMakeFiles/duckdb_planner_filter.dir/ub_duckdb_planner_filter.cpp.o planner/operator/CMakeFiles/duckdb_planner_operator.dir/ub_duckdb_planner_operator.cpp.o planner/subquery/CMakeFiles/duckdb_planner_subquery.dir/ub_duckdb_planner_subquery.cpp.o planner/CMakeFiles/duckdb_planner.dir/ub_duckdb_planner.cpp.o parser/constraints/CMakeFiles/duckdb_constraints.dir/ub_duckdb_constraints.cpp.o parser/expression/CMakeFiles/duckdb_expression.dir/ub_duckdb_expression.cpp.o parser/parsed_data/CMakeFiles/duckdb_parsed_data.dir/ub_duckdb_parsed_data.cpp.o parser/query_node/CMakeFiles/duckdb_query_node.dir/ub_duckdb_query_node.cpp.o parser/statement/CMakeFiles/duckdb_statement.dir/ub_duckdb_statement.cpp.o parser/tableref/CMakeFiles/duckdb_parser_tableref.dir/ub_duckdb_parser_tableref.cpp.o parser/transform/constraint/CMakeFiles/duckdb_transformer_constraint.dir/ub_duckdb_transformer_constraint.cpp.o parser/transform/expression/CMakeFiles/duckdb_transformer_expression.dir/ub_duckdb_transformer_expression.cpp.o parser/transform/helpers/CMakeFiles/duckdb_transformer_helpers.dir/ub_duckdb_transformer_helpers.cpp.o parser/transform/statement/CMakeFiles/duckdb_transformer_statement.dir/ub_duckdb_transformer_statement.cpp.o parser/transform/tableref/CMakeFiles/duckdb_transformer_tableref.dir/ub_duckdb_transformer_tableref.cpp.o parser/CMakeFiles/duckdb_parser.dir/ub_duckdb_parser.cpp.o function/aggregate/algebraic/CMakeFiles/duckdb_aggr_algebraic.dir/ub_duckdb_aggr_algebraic.cpp.o function/aggregate/distributive/CMakeFiles/duckdb_aggr_distr.dir/ub_duckdb_aggr_distr.cpp.o function/aggregate/holistic/CMakeFiles/duckdb_aggr_holistic.dir/ub_duckdb_aggr_holistic.cpp.o function/aggregate/nested/CMakeFiles/duckdb_aggr_nested.dir/ub_duckdb_aggr_nested.cpp.o function/aggregate/regression/CMakeFiles/duckdb_aggr_regr.dir/ub_duckdb_aggr_regr.cpp.o function/aggregate/CMakeFiles/duckdb_func_aggr.dir/ub_duckdb_func_aggr.cpp.o function/pragma/CMakeFiles/duckdb_func_pragma.dir/ub_duckdb_func_pragma.cpp.o function/scalar/blob/CMakeFiles/duckdb_func_blob.dir/ub_duckdb_func_blob.cpp.o function/scalar/date/CMakeFiles/duckdb_func_date.dir/ub_duckdb_func_date.cpp.o function/scalar/generic/CMakeFiles/duckdb_func_generic.dir/ub_duckdb_func_generic.cpp.o function/scalar/math/CMakeFiles/duckdb_func_math.dir/ub_duckdb_func_math.cpp.o function/scalar/operators/CMakeFiles/duckdb_func_ops.dir/ub_duckdb_func_ops.cpp.o function/scalar/sequence/CMakeFiles/duckdb_func_seq.dir/ub_duckdb_func_seq.cpp.o function/scalar/string/CMakeFiles/duckdb_func_string.dir/ub_duckdb_func_string.cpp.o function/scalar/nested/map/CMakeFiles/duckdb_func_map_nested.dir/ub_duckdb_func_map_nested.cpp.o function/scalar/nested/CMakeFiles/duckdb_func_nested.dir/ub_duckdb_func_nested.cpp.o function/scalar/system/CMakeFiles/duckdb_func_system.dir/ub_duckdb_func_system.cpp.o function/scalar/CMakeFiles/duckdb_func_scalar.dir/ub_duckdb_func_scalar.cpp.o function/table/system/CMakeFiles/duckdb_table_func_system.dir/ub_duckdb_table_func_system.cpp.o function/table/version/CMakeFiles/duckdb_func_table_version.dir/ub_duckdb_func_table_version.cpp.o function/table/CMakeFiles/duckdb_func_table.dir/ub_duckdb_func_table.cpp.o function/CMakeFiles/duckdb_function.dir/ub_duckdb_function.cpp.o catalog/catalog_entry/CMakeFiles/duckdb_catalog_entries.dir/ub_duckdb_catalog_entries.cpp.o catalog/default/CMakeFiles/duckdb_catalog_default_entries.dir/ub_duckdb_catalog_default_entries.cpp.o catalog/CMakeFiles/duckdb_catalog.dir/ub_duckdb_catalog.cpp.o common/crypto/CMakeFiles/duckdb_common_crypto.dir/ub_duckdb_common_crypto.cpp.o common/enums/CMakeFiles/duckdb_common_enums.dir/ub_duckdb_common_enums.cpp.o common/operator/CMakeFiles/duckdb_common_operators.dir/ub_duckdb_common_operators.cpp.o common/row_operations/CMakeFiles/duckdb_row_operations.dir/ub_duckdb_row_operations.cpp.o common/serializer/CMakeFiles/duckdb_common_serializer.dir/ub_duckdb_common_serializer.cpp.o common/types/CMakeFiles/duckdb_common_types.dir/ub_duckdb_common_types.cpp.o common/value_operations/CMakeFiles/duckdb_value_operations.dir/ub_duckdb_value_operations.cpp.o common/vector_operations/CMakeFiles/duckdb_vector_operations.dir/ub_duckdb_vector_operations.cpp.o common/CMakeFiles/duckdb_common.dir/ub_duckdb_common.cpp.o execution/expression_executor/CMakeFiles/duckdb_expression_executor.dir/ub_duckdb_expression_executor.cpp.o execution/merge_join/CMakeFiles/duckdb_merge_join.dir/ub_duckdb_merge_join.cpp.o execution/nested_loop_join/CMakeFiles/duckdb_nested_loop_join.dir/ub_duckdb_nested_loop_join.cpp.o execution/operator/aggregate/CMakeFiles/duckdb_operator_aggregate.dir/ub_duckdb_operator_aggregate.cpp.o execution/operator/filter/CMakeFiles/duckdb_operator_filter.dir/ub_duckdb_operator_filter.cpp.o execution/operator/helper/CMakeFiles/duckdb_operator_helper.dir/ub_duckdb_operator_helper.cpp.o execution/operator/join/CMakeFiles/duckdb_operator_join.dir/ub_duckdb_operator_join.cpp.o execution/operator/order/CMakeFiles/duckdb_operator_order.dir/ub_duckdb_operator_order.cpp.o execution/operator/persistent/CMakeFiles/duckdb_operator_persistent.dir/ub_duckdb_operator_persistent.cpp.o execution/operator/projection/CMakeFiles/duckdb_operator_projection.dir/ub_duckdb_operator_projection.cpp.o execution/operator/scan/CMakeFiles/duckdb_operator_scan.dir/ub_duckdb_operator_scan.cpp.o execution/operator/schema/CMakeFiles/duckdb_operator_schema.dir/ub_duckdb_operator_schema.cpp.o execution/operator/set/CMakeFiles/duckdb_operator_set.dir/ub_duckdb_operator_set.cpp.o execution/physical_plan/CMakeFiles/duckdb_physical_plan.dir/ub_duckdb_physical_plan.cpp.o execution/index/art/CMakeFiles/duckdb_art_index_execution.dir/ub_duckdb_art_index_execution.cpp.o execution/CMakeFiles/duckdb_execution.dir/ub_duckdb_execution.cpp.o main/relation/CMakeFiles/duckdb_main_relation.dir/ub_duckdb_main_relation.cpp.o main/CMakeFiles/duckdb_main.dir/ub_duckdb_main.cpp.o parallel/CMakeFiles/duckdb_parallel.dir/ub_duckdb_parallel.cpp.o storage/buffer/CMakeFiles/duckdb_storage_buffer.dir/ub_duckdb_storage_buffer.cpp.o storage/checkpoint/CMakeFiles/duckdb_storage_checkpoint.dir/ub_duckdb_storage_checkpoint.cpp.o storage/statistics/CMakeFiles/duckdb_storage_statistics.dir/ub_duckdb_storage_statistics.cpp.o storage/table/CMakeFiles/duckdb_storage_table.dir/ub_duckdb_storage_table.cpp.o storage/CMakeFiles/duckdb_storage.dir/ub_duckdb_storage.cpp.o transaction/CMakeFiles/duckdb_transaction.dir/ub_duckdb_transaction.cpp.o  -ldl ../third_party/fmt/libfmt.a ../third_party/libpg_query/libpg_query.a ../third_party/re2/libduckdb_re2.a ../third_party/miniz/libminiz.a ../third_party/utf8proc/libutf8proc.a ../third_party/hyperloglog/libhyperloglog.a -lpthread 
```
Note that the amalgamation build is missing `-ldl`, which imports the missing symbols into `libduckdb.so`. This is confirmed via `ldd`:
```
$ ldd build/*/src/libduckdb.so
build/amalgamation/src/libduckdb.so:
	linux-vdso.so.1 (0x00007ffebd5f7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7b31586000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f7b31402000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f7b3127f000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7b31265000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7b310a4000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f7b323f5000)
build/fullsrc/src/libduckdb.so:
	linux-vdso.so.1 (0x00007ffd39f63000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007efcf0292000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007efcf0271000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007efcf00ed000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007efceff6a000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007efceff50000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007efcefd8f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007efcf112b000)
```

**What should happen?**
Amalgamation builds should link the same system libraries as full-source builds.

**To Reproduce**
Steps to reproduce the behavior. Bonus points if those are only SQL queries.
```bash
mkdir -p build/amalgamation
python3 scripts/amalgamation.py
cd build/amalgamation
cmake ../.. -DAMALGAMATION_BUILD=ON
make
```

**Environment (please complete the following information):**
 - OS: Debian 10
 - DuckDB Version: 0.2.7 and HEAD

**Before submitting**
- [x] Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?
- [x] Have you tried this on the latest `master` branch? In case you cannot compile, you may find some binaries here: https://github.com/duckdb/duckdb/releases/tag/master-builds

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -32,10 +32,12 @@ if(AMALGAMATION_BUILD)
   endif()
 
   add_library(duckdb SHARED "${PROJECT_SOURCE_DIR}/src/amalgamation/duckdb.cpp")
+  target_link_libraries(duckdb ${CMAKE_DL_LIBS})
+  link_threads(duckdb)
+
   add_library(duckdb_static STATIC
               "${PROJECT_SOURCE_DIR}/src/amalgamation/duckdb.cpp")
-
-  link_threads(duckdb)
+  target_link_libraries(duckdb_static ${CMAKE_DL_LIBS})
   link_threads(duckdb_static)
 
   install(FILES "${PROJECT_SOURCE_DIR}/src/amalgamation/duckdb.hpp"

